text/0000-scoped-threads.md

```diff
@@ -0,0 +1,279 @@
+
+The alternative is to keep scoped threads in external crates. However, there are
+several advantages to having them in the standard library:
+
+* This is a very common and useful utility and is great for learning, testing, and exploratory
+  programming. Every person learning Rust will at some point encounter interaction
+  of borrowing and threads. There's a very important lesson to be taught that threads
+  *can* in fact borrow local variables, but the standard library doesn't reflect this.
+
+* Some might argue we should discourage using threads altogether and point people to
+  executors like Rayon and Tokio instead. But still,
+  the fact that `thread::spawn()` requires `F: 'static` and there's no way around it
+  feels like a missing piece in the standard library.
+
+* Implementing scoped threads is very tricky to get right so it's good to have a
+  reliable solution provided by the standard library.
+
+* There are many examples in the official documentation and books that could be
+  simplified by scoped threads.
+
+* Scoped threads are typically a better default than `thread::spawn()` because
+  they make sure spawned threads are joined and don't get accidentally "leaked".
+  This is sometimes a problem in unit tests, where "dangling" threads can accumulate
+  if unit tests spawn threads and forget to join them.
+
+* Users keep asking for scoped threads on IRC and forums
+  all the time. Having them as a "blessed" pattern in `std::thread` would be beneficial
+  to everyone.
+
+# Prior art
+[prior-art]: #prior-art
+
+Crossbeam has had
+[scoped threads](https://docs.rs/crossbeam/0.7.1/crossbeam/thread/index.html)
+since Rust 1.0.
+
+There are two designs Crossbeam's scoped threads went through. The old one is from
+the time `thread::scoped()` got removed and we wanted a sound alternative for the
+Rust 1.0 era. The new one is from the last year's big revamp:
+
+* Old: https://docs.rs/crossbeam/0.2.12/crossbeam/fn.scope.html
+* New: https://docs.rs/crossbeam/0.7.1/crossbeam/fn.scope.html
+
+There are several differences between old and new scoped threads:
+
+1. `scope()` now propagates unhandled panics from child threads.
+    In the old design, panics were silently ignored.
+    Users can still handle panics by manually working with `ScopedJoinHandle`s.
+
+2. The closure passed to `Scope::spawn()` now takes a `&Scope<'env>` argument that
+   allows one to spawn nested threads, which was not possible with the old design.
+   Rayon similarly passes a reference to child tasks.
+
+3. We removed `Scope::defer()` because it is not really useful, had bugs, and had
+   non-obvious behavior.
+
+4. `ScopedJoinHandle` got parametrized over `'scope` in order to prevent it from
+   escaping the scope.
+
+Rayon also has [scopes](https://docs.rs/rayon/1.0.3/rayon/struct.Scope.html),
+but they work on a different abstraction level - Rayon spawns tasks rather than
+threads. Its API is the same as the one proposed in this RFC.
+
+# Unresolved questions
```

Comment:
I'd like to raise a point on naming here.

#### The ambiguity of "Scope"

I remember the first time I learned about "scoped threads" and being confused about what they do. I remember wondering whether it referred to:
1. A "scope" as in "function scope"
2. A "scope" as in "telemetry scope"
3. Some other kind of "scope" specific to concurrent programming I was not yet aware of.

To this day I'm still not entirely clear on its origin. Though it seems likely that when this API was first introduced in 2014 (https://github.com/rust-lang/rfcs/pull/461) the name is a reference to [Boost C++ Scoped Threads](https://www.boost.org/doc/libs/1_53_0/doc/html/thread/ScopedThreads.html), which roughly seems to use "scope" as an analog for "grouping of"

#### Alternative Naming
 
Instead I would prefer we follow [Swift's example](https://github.com/apple/swift-evolution/blob/main/proposals/0304-structured-concurrency.md#task-groups), using simpler naming, and go with `ThreadGroup`:

```diff
- struct Scope<'env> {}
+ struct ThreadGroup<'env> {}

- fn scope<'env, F, T>(f: F) -> T {}
+ impl ThreadGroup<'env> {
+     fn new<'env, F, T>(f: F) -> T {}
+ }

- struct ScopedJoinHandle<'scope, T> {}
+ struct GroupJoinHandle<'scope, T> {}
```

The name `ThreadGroup` implies a type containing threads which work as a group — which is exactly what this API does. Seeing `ThreadGroup` referenced inside code should also be immediately clear as to what it does. If we contrast the two APIs:

```rust
// Immediately clear that this refers to a group of threads.
struct Thing {
    group: ThreadGroup,
}

// `Scope` requires additional context to clarify what it does.
struct Thing {
    scope: Scope,
}

Reply:
Scope means "lexical scope" (where "function scope" is one kind of lexical scope). They're named "scoped threads" because the API ensures the threads have exited before the scope ends.

A "thread group" is a much weaker concept - simply a set of threads that could be "joined" as one. A thread group need not be scoped.